src/client/datascience/kernel-launcher/remoteKernelFinder.ts

```diff
@@ -0,0 +1,160 @@
+// This class searches for a kernel that matches the given kernel name.
+// First it searches on a global persistent state, then on the installed python interpreters,
+
+import { Kernel } from '@jupyterlab/services';
+import { nbformat } from '@jupyterlab/coreutils';
+import { injectable, inject } from 'inversify';
+import { CancellationToken } from 'vscode';
+import { IDisposableRegistry, Resource } from '../../common/types';
+import { traceDecorators } from '../../logging';
+import { PythonEnvironment } from '../../pythonEnvironments/info';
+import { captureTelemetry } from '../../telemetry';
+import { Telemetry } from '../constants';
+import { findPreferredKernelIndex } from '../jupyter/kernels/helpers';
+import {
+    KernelConnectionMetadata,
+    DefaultKernelConnectionMetadata,
+    LiveKernelConnectionMetadata
+} from '../jupyter/kernels/types';
+import { PreferredRemoteKernelIdProvider } from '../notebookStorage/preferredRemoteKernelIdProvider';
+import {
+    IJupyterKernelSpec,
+    IJupyterSessionManager,
+    IJupyterSessionManagerFactory,
+    INotebookProviderConnection
+} from '../types';
+import { isInterpreter } from './localKernelFinder';
+import { IRemoteKernelFinder } from './types';
+
+// and finally on the default locations that jupyter installs kernels on.
+@injectable()
+export class RemoteKernelFinder implements IRemoteKernelFinder {
+    /**
+     * List of ids of kernels that should be hidden from the kernel picker.
+     */
+    private readonly kernelIdsToHide = new Set<string>();
+    constructor(
+        @inject(IDisposableRegistry) disposableRegistry: IDisposableRegistry,
+        @inject(PreferredRemoteKernelIdProvider)
+        private readonly preferredRemoteKernelIdProvider: PreferredRemoteKernelIdProvider,
+        @inject(IJupyterSessionManagerFactory) private jupyterSessionManagerFactory: IJupyterSessionManagerFactory
+    ) {
+        disposableRegistry.push(
+            this.jupyterSessionManagerFactory.onRestartSessionCreated(this.addKernelToIgnoreList.bind(this))
+        );
+        disposableRegistry.push(
+            this.jupyterSessionManagerFactory.onRestartSessionUsed(this.removeKernelFromIgnoreList.bind(this))
+        );
+    }
+    @traceDecorators.verbose('Find remote kernel spec')
+    @captureTelemetry(Telemetry.KernelFinderPerf)
+    public async findKernel(
+        resource: Resource,
+        connInfo: INotebookProviderConnection | undefined,
+        option?: nbformat.INotebookMetadata | PythonEnvironment,
+        _cancelToken?: CancellationToken
+    ): Promise<KernelConnectionMetadata | undefined> {
+        // Get list of all of the specs
+        const kernels = await this.listKernels(resource, connInfo);
```

Comment:
Would be good to pass in the `cancelToken` into this.

Reply:
Yeah good idea